File: 433 Minimum Genetic Mutation/Solution.cpp
class Solution {
public:
    int minMutation(string start, string end, vector<string>& bank) {
        queue<string> queue;
        queue.push(start);
        
        unordered_set<string> set;
        set.insert(start);
        int steps = 0;
        
        while (!queue.empty()) {
            int size = queue.size();
            while (size-- > 0) {
                string gene = std::move(queue.front());
                queue.pop();
                if (gene == end) return steps;
                for (const auto& str : bank) {
                    if (set.count(str) || !ValidMutation(gene, str)) continue;
                    queue.push(str);
                    set.insert(gene);
                }
                
            }
            ++steps;
        }
        
        return -1;
    }
    
private:
    bool ValidMutation(const string& s1, const string& s2) {
        int count = 0;
        for (int i = 0; i < s1.length(); ++i)
            if (s1[i] != s2[i] && count++) return false;
        return true;
    }
};

File: 433 Minimum Genetic Mutation/Solution.java
class Solution {
    public int minMutation(String start, String end, String[] bank) {
        if (start.equals(end)) return -1;
        Queue<String> queue = new LinkedList<>();
        queue.add(start);
        
        Set<String> set = new HashSet<>();
        int steps = 0;
        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                String cur = queue.poll();
                if (end.equals(cur)) return steps;
                for (String next : bank) {
                    if (!set.contains(next) && valid(cur, next)) {
                        set.add(next);
                        queue.add(next);
                    }
                }
            }
            steps++;
        }
        return -1;
    }
    
    boolean valid(String s1, String s2) {
        int cnt = 0;
        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) cnt++;
            if (cnt > 1) return false;
        }
        return true;
    }
}